src/server/managers/AppExternalComponentManager.ts

```diff
@@ -0,0 +1,74 @@
+import { IExternalComponent } from '../../definition/externalComponent';
+
+/**
+ * The external component manager for the Apps.
+ *
+ * An App will register external component during their `initialize` method.
+ * Then once an App's `onEnable` is called and it returns true,
+ * only then will that App's external component be enabled.
+ */
+export class AppExternalComponentManager {
+    /**
+     * The map that maintain the list of all registered components.
+     */
+    private providedComponents: Map<string, IExternalComponent>;
+    /**
+     * Contains the apps and the external components they have touhed.
+     * The key is appId and the value is external component.
+     * Doesn't matter whether the app provided, modified, disabled,
+     * or enabled. As long as an app touched an external component, then
+     * it is listed here.
+     */
+    private appTouchedComponents: Map<string, IExternalComponent>;
+    constructor() {
+        this.providedComponents = new Map<string, IExternalComponent>();
+        this.appTouchedComponents = new Map<string, IExternalComponent>();
+    }
+    /**
+     * Add the external component to the appTouchedComponents.
+     * @param appId the id of the app
+     * @param externalComponent the external component to register
+     * @param addToProvidedComponents whether also add the external
+     * component to the providedComponents, the value is false by default.
+     */
+    public registerComponent(appId: string, externalComponent: IExternalComponent, addToProvidedComponents: boolean = false): void {
+        this.appTouchedComponents.set(appId, externalComponent);
+
+        if (addToProvidedComponents) {
+            this.providedComponents.set(appId, externalComponent);
+        }
+    }
+    /**
+     * Remove the external compoent from the providedComponents by appId.
+     * @param appId the id of the app
+     */
+    public unregisterComponent(appId: string): void {
+        if (this.providedComponents.has(appId)) {
+            this.providedComponents.delete(appId);
+        }
+    }
+    /**
+     * Get the external component by the appId.
+     * @param appId the id of the app
+     */
+    public getExternalComponent(appId: string): IExternalComponent {
```

Comment:
I'm confused. Based upon the naming convention of nearly all of the methods in this class, an app can only have one external component. Yet the variable names inside here say otherwise. Can we update the method names to reflect what is happening? And if it is the case that only one external component is allowed per app, then we need to allow more than one per App before this gets merged. Unless there is a bigger technical limitation as to why we are limited to only one per app.

Reply:
Hi @graywolf336. Sorry for the confusion brought to you.

> an app can only have one external component.
yeah, it's true. We only allow one external component for one app currently. It's much easier for a Rocket.Chat administrator to enable/disable an external component by just enabling the app or disabling/uninstalling an app. It will make the problem become more complex if we support that one app owns multiple external components (We will need to add more control options in the apps admin panel for administrators to control).

> yet the variable names inside here say otherwise. Can we update the method names to reflect what is happening? 
the class AppExternalComponentManager maintains all the external components from all the apps installed in the Rocket.Chat. As I said before - one app corresponding to one external component, so when an administrator enabled/disabled an app, we need to enable/disable its external component correspondingly (here we need the method getExternalComponent, which returns an external component by the appId).

Reply:
> It's much easier for a Rocket.Chat administrator to enable/disable an external component by just enabling the app or disabling/uninstalling an app. It will make the problem become more complex if we support that one app owns multiple external components (We will need to add more control options in the apps admin panel for administrators to control).

I'm not so sure about this. We may only allow one external component per app now if it is technically much more difficult to implement, but as I stated in another comment, I see valid use cases for an app registering more than one external component

Reply:
OK. Will try to support multiple external components per app. @graywolf336 sorry for the confusion caused to you.

Reply:
resolved in the PR: https://github.com/RocketChat/Rocket.Chat.Apps-engine/pull/150

Reply:
No confusion here, I just wanted to bring the discussion up to understand the reasoning behind it. And that’s what this pull request is for, to discuss and improve. 🙂

We will need to list the external components in the configuration page of the App inside Rocket.Chat. That way an administrator can easily enable or disable what they want. We will also need a way for an app to tell what the status of a component is.